src/libstd/sys/windows/stdio.rs

```diff
@@ -1,220 +1,283 @@
-                         utf16.len() as u32,
+                         data.as_ptr() as c::LPCVOID,
+                         data.len() as u32,
                          &mut written,
                          ptr::null_mut())
     })?;
-
-    // FIXME if this only partially writes the utf16 buffer then we need to
-    //       figure out how many bytes of `data` were actually written
-    assert_eq!(written as usize, utf16.len());
-    Ok(utf8.len())
+    Ok(written as usize)
 }
 
 impl Stdin {
     pub fn new() -> io::Result<Stdin> {
-        Ok(Stdin {
-            utf8: Mutex::new(Cursor::new(Vec::new())),
-        })
-    }
-
-    pub fn read(&self, buf: &mut [u8]) -> io::Result<usize> {
-        let handle = match get(c::STD_INPUT_HANDLE)? {
-            Output::Console(c) => c,
-            Output::Pipe(p) => {
-                let handle = Handle::new(p);
-                let ret = handle.read(buf);
-                handle.into_raw();
-                return ret
-            }
-        };
-        let mut utf8 = self.utf8.lock().unwrap();
-        // Read more if the buffer is empty
-        if utf8.position() as usize == utf8.get_ref().len() {
-            let mut utf16 = vec![0u16; 0x1000];
-            let mut num = 0;
-            let mut input_control = readconsole_input_control(CTRL_Z_MASK);
-            cvt(unsafe {
-                c::ReadConsoleW(handle,
-                                utf16.as_mut_ptr() as c::LPVOID,
-                                utf16.len() as u32,
-                                &mut num,
-                                &mut input_control as c::PCONSOLE_READCONSOLE_CONTROL)
-            })?;
-            utf16.truncate(num as usize);
-            // FIXME: what to do about this data that has already been read?
-            let mut data = match String::from_utf16(&utf16) {
-                Ok(utf8) => utf8.into_bytes(),
-                Err(..) => return Err(invalid_encoding()),
-            };
-            if let Some(&last_byte) = data.last() {
-                if last_byte == CTRL_Z {
-                    data.pop();
-                }
-            }
-            *utf8 = Cursor::new(data);
-        }
-
-        // MemReader shouldn't error here since we just filled it
-        utf8.read(buf)
+        Ok(Stdin { surrogate: 0 })
     }
 }
 
-#[unstable(reason = "not public", issue = "0", feature = "fd_read")]
-impl<'a> Read for &'a Stdin {
+impl io::Read for Stdin {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        (**self).read(buf)
+        let handle = get_handle(c::STD_INPUT_HANDLE)?;
+        if !is_console(handle) {
+            let handle = Handle::new(handle);
+            let ret = handle.read(buf);
+            handle.into_raw(); // Don't close the handle
+            return ret;
+        }
+
+        if buf.len() == 0 {
+            return Ok(0);
+        } else if buf.len() < 4 {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput,
```

Comment:
FWIW this seems somewhat bad in the sense that this should handle arbitrary sized buffers, but I'm not too worried about this as it seems like this is vanishingly rare to come up in practice (e.g. libstd doesn't actually expose a way to do it due to its buffered reads I think), and we can always solve it with a 4-byte buffer in `Stdin` and some extra logic down the road.

Reply:
I thought about this case if `StdinRaw` is ever exposed. `Read::read_exact` continues to give a smaller slice of the buffer until it is completely filled, and is just about guaranteed to hit this error. But I can't imagine any reasonable use for `read_exact` on an unbuffered stdin.

> and we can always solve it with a 4-byte buffer in `Stdin` and some extra logic down the road.

Seems like a solution when unbuffered reads get exposed. But feels a bit to me like saying: you want to have unbuffered access to stdin? Here it is, and with a tiny buffer to make sure you can do arbitrary small reads. Not sure how to say it, it feels more like the concern of a wrapper.

Reply:
Yeah sounds good to me. If we leave this we'll just have to be sure to document it, but I think we can avoid that for now.